pypeit/utils.py

```diff
@@ -24,6 +24,170 @@
+        midpoint (:obj:`float`, optional):
+            Force the midpoint of the range to be centered on this
+            value. Default is the sample median.
+        default (:obj:`list`, optional):
+            Default limits to return if `a` has no data.
+
+    Returns:
+        :obj:`list`: Lower and upper boundaries for the data in `a`.
+    """
+    # Get the values to plot
+    _a = a.compressed() if isinstance(a, np.ma.MaskedArray) else np.asarray(a).ravel()
+    if len(_a) == 0:
+        # No data so return the default range
+        return default
+
+    # Set the starting and ending values based on a fraction of the
+    # growth
+    _lim = 1.0 if lim > 1.0 else lim
+    start, end = (len(_a)*(1.0+_lim*np.array([-1,1]))/2).astype(int)
+    if end == len(_a):
+        end -= 1
+
+    # Set the full range and multiply it by the provided factor
+    srt = np.ma.argsort(_a)
+    Da = (_a[srt[end]] - _a[srt[start]])*fac
+
+    # Set the midpoint
+    mid = _a[srt[len(_a)//2]] if midpoint is None else midpoint
+
+    # Return the range centered on the midpoint
+    return [ mid - Da/2, mid + Da/2 ]
+
+
+def nearest_unmasked(arr, use_indices=False):
+    """
+    Return the indices of the nearest unmasked element in a vector.
+
+    .. warning::
+        The function *uses the values of the masked data* for masked
+        elements. This means that if you want to know the nearest
+        unmasked element to one of the *masked* elements, the `data`
+        attribute of the provided array should have meaningful values
+        for these masked elements.
+
+    Args:
+        arr (`numpy.ma.MaskedArray`_):
+            Array to analyze. Must be 1D.
+        use_indices (:obj:`bool`, optional):
+            The proximity of each element in the array is based on
+            the difference in the array `data` values. Setting
+            `use_indices` to `True` instead bases the calculation on
+            the proximity of the element indices; i.e., find the
+            index of the nearest unmasked element.
+
+    Returns:
+        `numpy.ndarray`_: Integer array with the indices of the
+        nearest array elements, the definition of which depends on
+        `use_indices`.
+    """
+    # Check the input
+    if not isinstance(arr, np.ma.MaskedArray):
+        raise TypeError('Must provide a numpy masked array.')
+    if arr.ndim != 1:
+        raise ValueError('Must be a 1D array.')
+    if use_indices:
+        return nearest_unmasked(np.ma.MaskedArray(np.arange(arr.size), mask=arr.mask.copy()))
+
+    # Get the difference of each element with every other element
+    nearest = np.absolute(arr[None,:]-arr.data[:,None])
+    # Ignore the diagonal
+    nearest[np.diag_indices(arr.size)] = np.ma.masked
+    # Return the location of the minimum value ignoring the masked values
+    return np.ma.argmin(nearest, axis=1)
+
+
+def boxcar_smooth_rows(img, nave, wgt=None, mode='nearest'):
+    """
+    Boxcar smooth an image along rows.
```

Comment:
This is a general function.  I've added a note.

Reply:
Yeah I wrote this function (and all the other ones you've tightened up) so I understand how general it is:) I don't see a note addressing my comment in the docs. I don't see the harm in adding a line to the docs saying for "PypeIt spectral images, rows = spectral direction." Row and Column is ambiguous in python given the way images are stored. Is row the first index of the array or the second? Who knows. Python stores images one way but image viewers display them a different way. Why not just  don't tell the user in the docs what spectral and spatial are for PypeIt. It saves users a ton of time. It was a nightmare to port this code from IDL because everything was flipped relative to python, and I didn't have the documentation that I needed.